perf/src/recycler.rs

```diff
@@ -11,36 +18,128 @@ struct RecyclerStats {
     max_gc: AtomicUsize,
 }
 
-#[derive(Clone, Default)]
-pub struct Recycler<T> {
+#[derive(Debug, Default)]
+struct RecyclerShrinkStats {
+    resulting_size: u32,
+    target_size: u32,
+    ideal_num_to_remove: u32,
+    shrink_elapsed: u64,
+    drop_elapsed: u64,
+}
+
+impl RecyclerShrinkStats {
+    fn report(&self, shrink_metric_name: &'static str) {
+        datapoint_info!(
+            shrink_metric_name,
+            ("target_size", self.target_size as i64, i64),
+            ("resulting_size", self.resulting_size as i64, i64),
+            ("ideal_num_to_remove", self.ideal_num_to_remove as i64, i64),
+            ("recycler_shrink_elapsed", self.shrink_elapsed as i64, i64),
+            ("drop_elapsed", self.drop_elapsed as i64, i64)
+        );
+    }
+}
+
+#[derive(Clone)]
+pub struct Recycler<T: Reset> {
     recycler: Arc<RecyclerX<T>>,
+    shrink_metric_name: &'static str,
+}
+
+impl<T: Default + Reset> Recycler<T> {
+    pub fn new(shrink_metric_name: &'static str) -> Self {
+        Self {
+            recycler: Arc::new(RecyclerX::default()),
+            shrink_metric_name,
+        }
+    }
+
+    pub fn new_with_limit(shrink_metric_name: &'static str, limit: u32) -> Self {
+        Self {
+            recycler: Arc::new(RecyclerX::new(Some(limit))),
+            shrink_metric_name,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ObjectPool<T: Reset> {
+    object_pool: Vec<T>,
```

Comment:
nits: `object_pool.object_pool` in the other place of this code looks a bit uncool. How about renaming this to `objects`?

btw, I don't like `RecyclerX` as well, unrelated to your pr...